// adaptivecards-designer/lib/containers/host-container.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
var Adaptive = require("adaptivecards");
var hostConfig = require("../hostConfigs/sample.json");
var HostContainer = /** @class */ (function () {
    function HostContainer(name, styleSheet) {
        this.supportsActionBar = false;
        this.name = name;
        this.styleSheet = styleSheet;
        this._cardHost = document.createElement("div");
        this._cardHost.className = "cardHost";
    }
    HostContainer.prototype.initialize = function () {
        Adaptive.AdaptiveCard.elementTypeRegistry.reset();
        Adaptive.AdaptiveCard.actionTypeRegistry.reset();
        Adaptive.AdaptiveCard.useMarkdownInRadioButtonAndCheckbox = true;
        Adaptive.AdaptiveCard.useAdvancedCardBottomTruncation = false;
        Adaptive.AdaptiveCard.useAdvancedTextBlockTruncation = true;
    };
    HostContainer.prototype.getBackgroundColor = function () {
        return "#F6F6F6";
    };
    HostContainer.prototype.parseElement = function (element, json) {
        // Do nothing in base implementation
    };
    HostContainer.prototype.anchorClicked = function (element, anchor) {
        // Not handled by the host container by default
        return false;
    };
    HostContainer.prototype.getHostConfig = function () {
        return new Adaptive.HostConfig(hostConfig);
    };
    Object.defineProperty(HostContainer.prototype, "cardHost", {
        get: function () {
            return this._cardHost;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(HostContainer.prototype, "isFixedHeight", {
        get: function () {
            return false;
        },
        enumerable: true,
        configurable: true
    });
    return HostContainer;
}());
exports.HostContainer = HostContainer;
