// adaptivecards-designer/lib/containers/outlook/outlook-container.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
var Adaptive = require("adaptivecards");
var host_container_1 = require("../host-container");
var hostConfig = require("../../hostConfigs/outlook-desktop.json");
var OutlookContainer = /** @class */ (function (_super) {
    __extends(OutlookContainer, _super);
    function OutlookContainer() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    OutlookContainer.prototype.renderTo = function (hostElement) {
        hostElement.classList.add("outlook-frame");
        hostElement.appendChild(this.cardHost);
    };
    OutlookContainer.prototype.initialize = function () {
        _super.prototype.initialize.call(this);
        Adaptive.AdaptiveCard.actionTypeRegistry.unregisterType("Action.Submit");
        Adaptive.AdaptiveCard.actionTypeRegistry.registerType("Action.Http", function () { return new Adaptive.HttpAction(); });
        Adaptive.AdaptiveCard.useMarkdownInRadioButtonAndCheckbox = false;
    };
    OutlookContainer.prototype.parsePadding = function (json) {
        if (json) {
            if (typeof json === "string") {
                var uniformPadding = Adaptive.getEnumValue(Adaptive.Spacing, json, Adaptive.Spacing.None);
                return new Adaptive.PaddingDefinition(uniformPadding, uniformPadding, uniformPadding, uniformPadding);
            }
            else if (typeof json === "object") {
                return new Adaptive.PaddingDefinition(Adaptive.getEnumValue(Adaptive.Spacing, json["top"], Adaptive.Spacing.None), Adaptive.getEnumValue(Adaptive.Spacing, json["right"], Adaptive.Spacing.None), Adaptive.getEnumValue(Adaptive.Spacing, json["bottom"], Adaptive.Spacing.None), Adaptive.getEnumValue(Adaptive.Spacing, json["left"], Adaptive.Spacing.None));
            }
        }
        return null;
    };
    OutlookContainer.prototype.parseElement = function (element, json) {
        if (element instanceof Adaptive.AdaptiveCard) {
            var card = element;
            var actionArray = [];
            card["resources"] = { actions: actionArray };
            if (typeof json["resources"] === "object") {
                var actionResources = json["resources"]["actions"];
                for (var i = 0; i < actionResources.length; i++) {
                    var action = Adaptive.AdaptiveCard.actionTypeRegistry.createInstance(actionResources[i]["type"]);
                    if (action) {
                        action.parse(actionResources[i]);
                        action.setParent(card);
                        actionArray.push(action);
                    }
                }
            }
        }
        if (element instanceof Adaptive.Image) {
            element.backgroundColor = json["backgroundColor"];
        }
        if (element instanceof Adaptive.Container) {
            var padding = this.parsePadding(json["padding"]);
            if (padding) {
                element.padding = padding;
            }
        }
        if (element instanceof Adaptive.ColumnSet) {
            var padding = this.parsePadding(json["padding"]);
            if (padding) {
                element.padding = padding;
            }
        }
    };
    OutlookContainer.prototype.anchorClicked = function (element, anchor) {
        var regEx = /^action:([a-z0-9]+)$/ig;
        var rootCard = element.getRootElement();
        var matches = regEx.exec(anchor.href);
        if (matches) {
            var actionId = matches[1];
            if (rootCard) {
                var actionArray = rootCard["resources"]["actions"];
                for (var i = 0; i < actionArray.length; i++) {
                    if (actionArray[i].id == actionId) {
                        actionArray[i].execute();
                        return true;
                    }
                }
            }
        }
        return false;
    };
    OutlookContainer.prototype.getHostConfig = function () {
        return new Adaptive.HostConfig(hostConfig);
    };
    return OutlookContainer;
}(host_container_1.HostContainer));
exports.OutlookContainer = OutlookContainer;
